File: solution/384. Shuffle an Array/384.cpp
class Solution {
public:
    Solution(vector<int>& nums) {
        A_ = nums;
    }
    
    vector<int> reset() {
        return A_;
    }
    
    vector<int> shuffle() {
        int n = A_.size();
        vector<int> res = A_;
        for (int i = 0; i < n; ++i) {
            int j = rand() % (n - i) + i;
            swap(res[i], res[j]);
        }
        return res;
    }
    
private:
    vector<int> A_;
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(nums);
 * vector<int> param_1 = obj->reset();
 * vector<int> param_2 = obj->shuffle();
 */

File: solution/384. Shuffle an Array/384.java
class Solution {
    private int[] nums;
    
    public Solution(int[] nums) {
        this.nums = nums;
    }
    
    public int[] reset() {
        return nums;
    }
    
    public int[] shuffle() {
        int[] res = nums.clone();
        for (int i = 0; i < nums.length; ++i) {
            int rand = (int)(Math.random() * (nums.length - i)) + i;
            swap(res, i, rand);
        }
        return res;
    }
    
    private void swap(int[] res, int i, int j) {
        int tmp = res[i];
        res[i] = res[j];
        res[j] = tmp;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * int[] param_1 = obj.reset();
 * int[] param_2 = obj.shuffle();
 */
